### ISSUE TITLE
Container Registry

### ISSUE THREAD

**Issue by croumegous** [hardpixel/dokku-gitlab#2]
Hello, 

Do you have any idea how to create a container registry with dokku-gitlab.

I've tried several things but none is working great.
I've tried adding the following settings in /etc/gitlab/gitlab.rb: 
# 1
```bash
registry_external_url 'http://gitlab.registry.<domain>.com'
registry['enable'] = true
```
Added the domain in dokku and add the certificate with letsencrypt for this new domain.
When I do a `docker push` it start to work for a 0.1s then it show me the following error. (also the image container name is created in gitlab but is empty as it didn't finish to push)
```bash
 error parsing HTTP 400 response body: invalid character '<' looking for beginning of value: "<html>\r\n<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>400 Bad Request</h1></center>\r\n<center>The plain HTTP request was sent to HTTPS port</center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n"
```

# 2
```bash
registry_external_url 'http://gitlab.<domain>.com:5050' # same url as gitlab with an explicit port
registry['enable'] = true
```
However it does not work because it try to connect to <host_ip>:5050 and bypass dokku/gitlab container

# 3
Almost like 1st one but with https
```bash
registry_external_url 'https://gitlab.registry.<domain>.com'
registry['enable'] = true
```
after reconfigure gitlab frontpage was showing the basic nginx/dokku error. In the container logs it said :
```bash
==> /var/log/gitlab/nginx/error.log <==
2021/06/21 15:12:04 [emerg] 13232#0: cannot load certificate key "/etc/gitlab/ssl/gitlab.registry.<domain>.com.key": BIO_new_file() failed (SSL: error:02001002:system library:fopen:No such file or directory:fopen('/etc/gitlab/ssl/gitlab.registry.<domain>.com.key','r') error:2006D080:BIO routines:BIO_new_file:no such file)
```
So I added the dokku certificate content manually from `dokku certs:show gitlab key` into `/etc/gitlab/ssl/gitlab.registry.<domain>.com.key`  
And `dokku certs:show gitlab crt` into `/etc/gitlab/ssl/gitlab.registry.<domain>.com.crt`

However now on docker push I got 404 and indeed the url https://gitlab.registry.domain.com is now redirecting me to the regular gitlab frontpage instead of the registry.

\  
\  
\  

The thing that seems to work the most is the first solution as it seems to at least create something in the container registry.  
However i didn't find a solution for this http<->https error maybe a config with dokku, maybe with gitlab.rb ...

Do you have by any chance an idea on how to make it work ?
Thanks

**Comment by jonian** [comment]
Hi @croumegous, I haven't used a docker registry with gitlab on dokku, so I do not know how to make it work.

> However it does not work because it try to connect to <host_ip>:5050 and bypass dokku/gitlab container

Did you expose the container port?

```
dokku docker-options:add gitlab deploy '-p 5050:5050'
```

**Comment by jonian** [comment]
Documentation for docker-options [here](https://dokku.com/docs/advanced-usage/docker-options/)

**Comment by croumegous** [comment]
Thanks for the quick reply.  
I tried this solution and indeed it kind of work.  
I achieve to send an image to the registry, however I can't certified it with letsencrypt because of the explicit port.  
And I was always getting this error (on docker push or pull) : 
`x509: certificate signed by unknown authority`
So I needed to use [insecure registry](https://docs.docker.com/registry/insecure/) in order to push to the repository.  
But can't be used reliably for example in gitlab CI/CD to pull or push images.

So I guess I will give up using the registry with dokku-gitlab.  
Unless you have another bright idea, I think I will close this issue.
